setup/migrate.py

```diff
@@ -191,6 +191,62 @@ def migration_14(env):
 	db = os.path.join(env["STORAGE_ROOT"], 'mail/users.sqlite')
 	shell("check_call", ["sqlite3", db, "CREATE TABLE auto_aliases (id INTEGER PRIMARY KEY AUTOINCREMENT, source TEXT NOT NULL UNIQUE, destination TEXT NOT NULL, permitted_senders TEXT);"])
 
+def migration_15(env):
+	backup_configuration_file = os.path.join(env["STORAGE_ROOT"], 'backup', 'custom.yaml')
+	if os.path.exists(backup_configuration_file):
+		import yaml
+
+		with open(backup_configuration_file, 'r') as f:
+			custom_config = yaml.load(f, Loader=yaml.SafeLoader)
+		if not isinstance(custom_config, dict): raise ValueError()
+
+		# Converting the previous configuration (which was not very clear)
+		# into the new configuration format which also provides
+		# a "type" attribute to distinguish the type of backup.
+		scheme = custom_config["target"].split(":")[0]
+		if scheme == "off":
+			custom_config = {
+				"type": "off"
+			}
+		elif scheme == "file":
+			custom_config = {
+				"type": "local",
+				"min_age_in_days": custom_config["min_age_in_days"]
+			}
+		elif scheme == "rsync":
+			custom_config = {
+				"type": "rsync",
+				"target_url": custom_config["target"],
+				"min_age_in_days": custom_config["min_age_in_days"]
+			}
+		elif scheme == "s3":
+			import urllib.parse
+			url = urllib.parse.urlparse(custom_config["target"])
+			target_url = url.scheme + ":/" + url.path
+			s3_endpoint_url = "https://" + url.netloc
+			s3_access_key_id = custom_config["target_user"]
+			s3_secret_access_key = custom_config["target_pass"]
+			custom_config = {
+				"type": "s3",
+				"target_url": target_url,
+				"s3_endpoint_url": s3_endpoint_url,
+				"s3_access_key_id": custom_config["target_user"],
+				"s3_secret_access_key": custom_config["target_pass"],
+				"min_age_in_days": custom_config["min_age_in_days"]
+			}
+		elif scheme == "b2":
+			custom_config = {
+				"type": "b2",
+				"target_url": custom_config["target"],
+				"min_age_in_days": custom_config["min_age_in_days"]
+			}
+		else:
+			raise ValueError("Unexpected scheme during the conversion of the previous config to the new format.")
```

Comment:
See previous: we can recover gracefully-ish here by disabling backups and returning early.

Reply:
Solved with c2636f68f898fd77368350ae2039b08e47de33b3.